pkgs/applications/virtualization/xen/4.18/4.18.nix

```diff
@@ -0,0 +1,194 @@
+{
+  lib,
+  callPackage,
+  fetchpatch,
+  fetchgit,
+  fetchFromGitHub,
+  ocaml-ng,
+
+  withInternalQEMU ? true,
+
+  withInternalTraditionalQEMU ? true,
+
+  withInternalSeaBIOS ? true,
+  withSeaBIOS ? !withInternalSeaBIOS,
+  seabios,
+
+  withInternalOVMF ? true,
+  withOVMF ? !withInternalOVMF,
+  OVMF,
+
+  withInternalIPXE ? true,
+  withIPXE ? !withInternalIPXE,
+  ipxe,
+
+  # QEMU
+  udev,
+  pciutils,
+  xorg,
+  SDL,
+  pixman,
+  acl,
+  glusterfs,
+  spice-protocol,
+  usbredir,
+  alsa-lib,
+  glib,
+  python311,
+
+  # OVMF
+  nasm,
+  bash,
+  ...
+}@args:
+
+assert withInternalSeaBIOS -> !withSeaBIOS;
+assert withInternalOVMF -> !withOVMF;
+assert withInternalIPXE -> !withIPXE;
+
+let
+  upstreamPatches = import ../patches.nix { inherit fetchpatch; };
+
+  qemuDeps = [
+    udev
+    pciutils
+    xorg.libX11
+    SDL
+    pixman
+    acl
+    glusterfs
+    spice-protocol
+    usbredir
+    alsa-lib
+    glib
+    python311
+  ];
+in
+
+callPackage (import ../generic.nix rec {
+  branch = "4.18";
+  version = "4.18.2";
+  qemuVersion = "4.18.1";
+  seaBIOSVersion = "1.16.3";
+  ovmfVersion = "202305";
+
+  src = fetchgit {
+    url = "git://xenbits.xenproject.org/xen.git";
+    rev = "RELEASE-${version}";
+    hash = "sha256-pHCjj+Bcy4xQfB9xHU9fccFwVdP2DXrUhdszwGvrdmY=";
+  };
+
+  # Sources needed to build tools and firmwares.
+  prefetchedSources =
+    lib.attrsets.optionalAttrs withInternalQEMU {
+      qemu-xen = {
+        src = fetchgit {
+          url = "git://xenbits.xenproject.org/qemu-xen.git";
+          rev = "qemu-xen-${qemuVersion}";
+          hash = "sha256-BX+LXfNzwdUMALwwI1ZDW12dJ357oynjnrboLHREDGQ=";
+        };
+        buildInputs = qemuDeps;
+        postPatch = ''
+          substituteInPlace scripts/tracetool.py \
+            --replace-fail "/usr/bin/env python" "${python311}/bin/python"
+        '';
+        meta.description = "Xen's fork of upstream QEMU";
+      };
+    }
+    // lib.attrsets.optionalAttrs withInternalTraditionalQEMU {
```

Comment:
Unless you have a very important need to keep qemu-traditional, don't (and the "4.15 broken" starting point strongly suggests not).

It is explicitly unsupported in this this usecase

Reply:
`qemu-traditional` has been removed. Thanks!